---
title: "Self Quiz and Exercises for Week 6"
output: learnr::tutorial
runtime: shiny_prerendered
---

```{r setup, include=FALSE}
library(learnr)
knitr::opts_chunk$set(echo = FALSE)
```

## self-quiz: R data structures


```{r,eval=T,echo=T}
a_mat <- matrix(1:8, ncol=2)
a_mat
```


### Q1

```{r q1, echo=FALSE}
question("Based on the contents of a_mat, which of the following code would print out the value 8 if added below the existing lines in the code chunk above?",
  answer("a_mat[2]"),
  answer("a_mat[2,]"),
  answer("a_mat[4]"),
  answer("a_mat[2,4]"),
  answer("a_mat[4,2]",correct = TRUE),
  answer("None of them")
)
```


### Q2

```{r q2, echo=FALSE}
question("Based on the contents of a_mat, which of the following code would return a vector containing these values: (5,6,7,8)?",
  answer("a_mat[2]"),
  answer("a_mat[2,]"),
  answer("a_mat[,2]",correct = TRUE),
  answer("a_mat[2,1]"),
  answer("a_mat[4,2]"),
  answer("None of them")
)
```


### Q3

Data frames can be created using a syntax similar to what we used when creating a list with named elements. The code below creates a data frame with three columns, as shown in the printout below. 

```{r,eval=T,echo=T}
df <- data.frame("first"=c("Doctor","The","Doctor","Dr.","Dr."),
                 "second"=c("Dre","Doctor","Octopus","Morin","Evil"),
                 "power"=c("Sick beats","Time travel","Robotic tentacles","confusion","evil"),
                 "coolness"=c(9,12,3,8,4),
                 row.names=c("a","b","c","d","e"))
df
```

```{r q3,echo=FALSE}
question("Which of the following will return a data frame with the same contents but with the rows arranged from lowest to highest coolness score?",
  answer("df[order(df$coolness)]"),
  answer("df[order(df$coolness),]",correct = TRUE),
  answer("df[,order(df$coolness)]"),
  answer("df[order(rownames(df)),]")
)
```

### Q4

```{r q4, echo=FALSE}
question("Which code will produce a data frame containing only rows with the first name set to either Dr or Doctor?",
  answer("None of them"),
  answer("df[grepl('D',df$first)]"),
  answer("df[grepl('D',df$first),]",correct = TRUE),
  answer("df[grepl('Doctor',df$first),]")
)
```


```{r,echo=T,eval=T}

days <- c("Monday","Tuesday","Wednesday","Thursday","Friday","Saturday","Sunday")


```

### Q5

```{r q5, echo=FALSE}
question("Which line of code would subset the days vector to just weekdays?",
  answer("None of them"),
  answer("days[ grepl('S',days)]"),
  answer("! days[ grepl('S',days)]"),
  answer("days[ !grepl('S',days)]",correct = TRUE)
)
```

### Q6

```{r q6, echo=FALSE}
question("Which of the following statements comparing R data frames and matrixes are accurate?",
  answer("Matrix and data frames both enforce unique row and column names"),
  answer("Both can only contain data of a single type"),
  answer("Rows or columns can be subset directly into vectors using [] sytax and the corresponding row or column index"),
  answer("In data frames, columns can contain a mix of data types but matrixes do not allow this."),
  answer("None of the statements are accurate",correct = TRUE)
)
```


### Q7

```{r,echo=T}
x <- c(1,2,3)
y <- 2
```


```{r q7, echo=FALSE}
question("What is x+y?",
  answer("c(3,2,3)"),         
  answer("c(1,2,3,2)"),
  answer("c(3,4,5)",correct = TRUE)
)
```


### Q8

```{r,echo=T}
df <- data.frame(
  x = 1:3,
  y = 2
)
```


```{r q8, echo=FALSE}
question("What is df$y?",
  answer("2"),         
  answer("c(2, 2, 2)", correct=TRUE),
  answer("6"),
  answer("c(2,4,6)")
)
```


### Q9

```{r,q9,echo=FALSE}
question("Which code will return a vector containing the mean of each column in df?",
         answer("mean(df)"),
         answer("mean(df$x) + mean(df$y)"),
         answer("apply(df,2,mean)", correct=TRUE),
         answer("apply(mean, df)"),
         answer("df(mean$df)"),
         answer("apply(df,1,mean)"))
```

###Q10

```{r,q10code,exercise=TRUE}
# a nested list
movie_list <- list(
  list(title = "The Matrix",
       year = 1999,
       cast = c("Keanu Reeves", "Laurence Fishburne", "Carrie-Anne Moss", "Hugo Weaving", "Joe Pantoliano")),
  list(title = "The Matrix Resurrections",
       year = 2021,
       cast = c("Keanu Reeves", "Carrie-Anne Moss", "Jada Pinkett Smith", "Lambert Wilson"))
)
```

```{r,q10,echo=FALSE}
question("Which code will return just the cast of The Matrix Resurrections?",
         answer("movie_list$cast"),
         answer("movie_list[[2]]$cast",correct=TRUE),
         answer("movie_list[2]$cast"),
         answer("movie_list[[2$cast]]")
         )
```


### Getting vexxed by vectors

*Add to the code below to create a new vector from this_vec that contains the elements in numerical order. Your code should print the contents of your new vector. When you have the correct implementation, you should get a printout identical to what is shown immediately below this.*

```{r,echo=F,eval=T}
this_vec = c(8,22,12,40,3,1)
this_vec[order(this_vec)]
```

In your solution you must use the `order` function. 

```{r try1, exercise=TRUE, exercise.lines = 5}
this_vec = c(8,22,12,40,3,1)

```

*Add to the code below to create a new vector from this_vec that contains the elements in numerical order. When you have the correct implementation, you should get a printout identical to what is shown immediately below this.*


```{r try2, exercise=TRUE, exercise.lines = 5}
this_list = list(z="?",y="hard",m="really",c="now",f="was",w="so")

```

### Listing everything you like about lists

*Add to the code below to create a new vector from this_list that contains the elements in the order according to the names (not the values). Once you have that, print the vector contents. When you have the correct implementation, you should get a sensible sentence with proper punctuation (below the corresponding names from the original list).*

Hint/warning: Although `[[]]` is useful for getting a vector out of a list, you can't use it with vectors of length >1. 

```{r print-limit, exercise=TRUE, exercise.eval=TRUE}
this_list = list(h="that",z="?",y="difficult",m="really",c="Now",f="was",w="so")
```

```{r print-limit-hint}
order(names(this_list)) #here is the first step you will need. 

```

### Controlling The Matrix (like Neo)

*Add to the code below to order the columns of the matrix based on the numeric values in the second row (i.e. row index 2). Print out the ordered matrix when you're done. You should be able to see a secret message in the bottom two lines of your matrix when you have the correct answer. *

```{r reorder, exercise=TRUE, exercise.eval=TRUE}

alpha_mat = matrix(c("M","z","d","E","w","o","o","G","L","J","o","o","A","c","G","D"),nrow=4)
alpha_mat
```

```{r reorder-hint}
order(alpha_mat[2,]) #this should get you started
```

### Can't think of a way to frame a pun about data

*The code below loads the first 125,000 of Ryan's genotypes from 23andMe. Add some code below this to subset this data frame to create a new data frame containing only rows that are heterozygous (not homozygous). Be sure the data frame doesn't contain any "no call" genotypes, which look like this: "--". When you have the right answer, you should have a data frame with 59930 rows. Confirm this by running the dim() function on your new data frame. *

```{r df, exercise=TRUE, exercise.eval=TRUE}

morin = read.csv("~/MBB243/data/Morin_genotypes.txt",sep="\t",nrows = 125000)
table(morin$genotype)

```

## Week 6 Lab assignment

### Background

You have been provided with the coding DNA sequences (CDS) for the various isoforms of all human genes on chromosome 7. A snippit from the fasta file is shown below. You may notice that many of the sequence records will start with ATG (the standard start codon for translation). Not all of them will, for a few reasons. 

```
>ENSG00000128596|ENST00000487361|ENST00000487361.5|CCDC136
ATGCAAGCTATGGAGGGGGAGGTGTTACTCCCAGCTCTCTATGAGGAGGAAGAGGAAGAG
GAAGAGGAGGAAGAAGAGGTGGAAGAAGAAGAAGAACAAGTGCAGAAAGGTGGCAGTGTT
GGCTCTCTGTCAGTCAACAAGCACCGGGGACTGAGCCTCACGGAGACAGAGCTGGAGGAG
CTGCGGGCTCAGGTGCTGCAGCTGGTGGCAGAACTGGAGGAGACCCGGGAACTGGCAGGG
```

For the sake of this assignment, you should only process sequences that start with ATG and ignore the rest. By my count, there are 4374 sequences in this file that you will need to process. Some of these isoforms have no coding sequence at all so will instead have FASTA records that look like the example below. Your code will need to be able to deal with these (i.e. not crash when these are encountered). 

```
>ENSG00000122642|ENST00000464334|ENST00000464334.1|FKBP9
Sequence unavailable
```


The first goal of this assignment is for you to perform in silico translation of each sequence into the corresponding protein sequence using the genetic code. You should write this set of protein sequences to a new fasta file that you will name `translated.fasta`. This should be performed using a Python script that you develop based on the template provided. You can use a mix of BioPython functions, functions provided in previous labs, or custom Python code to accomplish this, as you see fit. 

As a reminder, here are the single-letter and three-letter codes for each of the 20 amino acids.

```
G		Glycine		Gly									P		Proline		Pro
A		Alanine		Ala									V		Valine		Val
L		Leucine		Leu									I		Isoleucine		Ile
M		Methionine		Met							    C		Cysteine		Cys
F		Phenylalanine		Phe						    Y		Tyrosine		Tyr
W		Tryptophan		Trp							    H		Histidine		His
K		Lysine		Lys									R		Arginine		Arg
Q		Glutamine		Gln								N		Asparagine		Asn
E		Glutamic Acid		Glu						    D		Aspartic Acid		Asp
S		Serine		Ser									T		Threonine		Thr

```

The second goal of this assignment is to tabulate the number of each amino acid individually in each of those protein sequences you've written to `translated.fasta`. In theory, it's not necessary to read the sequences back in if you prefer to perform this step in parallel with the previous step. One important thing to keep in mind is that you need to use a unique identifier for each sequence. I suggest you use the portion of the identifier that we used in an earlier example (it begins with ENST). You are expected to implement a new function `count_amino_acids` that will do the amino acid tabulation for you. Your script should output these counts to a tab-delimited text file with one row per transcript (i.e. per sequence record) and one column for each amino acid. The first column of this file should be the unique ID for your transcript. The output file should also have a header that defines what each column is. This file should be named `amino_count.tsv`. For amino acids, the header can be the single-letter code for that amino acid. A toy example of the process is shown below. Because there are 20 amino acids and 4374 protein-coding transcripts, your output file should have *4375* rows and *21* columns (including the row names and header). The file `example.tsv` in this directory shows the exact format. The template R script currently loads that file. You will need to modify your R script to use your own file as input. I suggest you implement your R code first, using this file to debug it, before you change your R code to use your own file as input. 


```
example translated proteins:

>ENST00001
MAPCWNLGFV
>ENST00002
MCPWARQQM

Example output file (spaces used here insted of tabs to help them fit on your screen):

ID    G P A V L I M C F Y W H K R Q N E D S T		
ENSG00001   1 1 1 1 1 0 1 1 1 0 1 0 0 0 0 1 0 0 0 0
ENSG00002   0 1 1 0 0 0 2 0 0 0 1 0 0 1 2 0 0 0 0 0

```


### Instructions

This week's assignment requires some coding in both Python and R. You have been provided with template scripts that you should make a copy of in the lab06 directory before you start. Replace "studentname" with your name. Reminder: To copy a file use the `cp` command e.g. `cp template-studentname.R template-Ryan.R`. The code in these templates has been provided to help you get started. Don't remove or modify any of the existing code except where clearly stated. 

```{bash,eval=T}
ls template*
```

The Python script should be modified to perform some analysis on the contents of human_CDS_chr7.fa, which is currently compressed. Be sure to use `gunzip` to decompress it before you start! In your Python code, you will need to do the following:

* Load the sequence of each transcript 
* Write the translated protein from those that begin with ATG to a file named `translated.fasta`
* Write the tabulated amino acid counts for each of those transcripts to a new file named `amino_count.tsv`, which will be in the same format as the example provided

After making a copy with your name in the file name, you should modify your R template to do the following with `amino_count.tsv`. The output of this script does not need to be a file. It can simply print to the terminal using a print statement. A print statement example is provided in the template. Modify it accordingly. 

* Read the tsv file into a data frame using the first column in amino_count.tsv as your row names
* Calculate the total number of times each of the 20 amino acids appear in the proteins in your file (based on this data frame)
* Determine which protein contains the most copies of a specific amino acid as specified in the table below. For example, Ryan's script should determine which translated sequence contains the most methionine (M) residues.


|Student|AA|
|-|-|
|Ryan|M|
|Adam|T|
|Alex|S|
|Elizabeth|D|
|Houman|E|
|Jonathan|N|
|Kelly|Q|
|Marina|Y|
|Rachel|I|
|Rebecca|K|
|Tanvi|C|


